src/service/logs/loki.rs

```diff
@@ -0,0 +1,608 @@
+    .await
+    .map_err(|e| {
+        log::error!("Error while writing logs: {}", e);
+        LokiError::from(anyhow::anyhow!("Multi-stream ingestion failed: {:?}", e))
+    })?;
+
+    log::info!("[LOKI] Successfully processed streams for org '{}'", org_id);
+    Ok(IngestionResponse::new(200, vec![]))
+}
+
+fn process_json_request(
+    request: LokiPushRequest,
+    json_data_by_stream: &mut HashMap<String, super::O2IngestJsonData>,
+) -> Result<(), LokiError> {
+    for loki_stream in request.streams {
+        let stream_name = determine_service_stream_name(&loki_stream.stream);
+
+        for entry in loki_stream.values {
+            let mut record = json::Map::new();
+
+            for (key, value) in &loki_stream.stream {
+                record.insert(key.clone(), json::Value::String(value.clone()));
+            }
+            if let Some(metadata) = entry.structured_metadata {
+                for (key, value) in metadata {
+                    record.insert(key, json::Value::String(value));
+                }
+            }
+
+            let timestamp_us =
+                entry
+                    .timestamp
+                    .parse::<i64>()
+                    .map_err(|e| LokiError::InvalidTimestamp {
+                        message: format!("Invalid timestamp '{}': {}", entry.timestamp, e),
+                    })?
+                    / 1_000;
+
+            record.insert("message".to_string(), json::Value::String(entry.line));
+            record.insert(
+                "_timestamp".to_string(),
+                json::Value::Number(timestamp_us.into()),
+            );
+
+            json_data_by_stream
+                .entry(stream_name.clone())
+                .or_insert((Vec::new(), Some(0)))
+                .0
+                .push((timestamp_us, record));
+        }
+    }
+    Ok(())
+}
+
+fn process_protobuf_request(
+    request: loki_rpc::PushRequest,
+    json_data_by_stream: &mut HashMap<String, super::O2IngestJsonData>,
+) -> Result<(), LokiError> {
+    for loki_stream in request.streams {
+        let labels = parse_prometheus_labels(&loki_stream.labels)?;
+
+        let stream_name = determine_service_stream_name(&labels);
+
+        for entry in loki_stream.entries {
+            let mut record = json::Map::new();
+
+            for (key, value) in &labels {
+                record.insert(key.clone(), json::Value::String(value.clone()));
+            }
+
+            let timestamp_us = if let Some(ts) = entry.timestamp {
+                let seconds_us = ts.seconds * 1_000_000;
+                let nanos_us = ts.nanos as i64 / 1_000;
+                seconds_us + nanos_us
+            } else {
+                chrono::Utc::now().timestamp_micros()
+            };
+
+            record.insert("message".to_string(), json::Value::String(entry.line));
+
+            for label_pair in entry.structured_metadata {
+                record.insert(label_pair.name, json::Value::String(label_pair.value));
```

Comment:
This section before handle `ts` and `message` in JSON section, but in Protobuf after it. suggest keep some logic.